cpp/src/arrow/extension/variable_shape_tensor.cc

```diff
@@ -0,0 +1,341 @@
+  const size_t ndim =
+      std::static_pointer_cast<FixedSizeListType>(storage_type->field(0)->type())
+          ->list_size();
+
+  rj::Document document;
+  if (document.Parse(serialized_data.data(), serialized_data.length()).HasParseError()) {
+    return Status::Invalid("Invalid serialized JSON data: ", serialized_data);
+  }
+
+  std::vector<int64_t> permutation;
+  if (document.HasMember("permutation")) {
+    for (auto& x : document["permutation"].GetArray()) {
+      permutation.emplace_back(x.GetInt64());
+    }
+    if (permutation.size() != ndim) {
+      return Status::Invalid("Invalid permutation");
+    }
+  }
+  std::vector<std::string> dim_names;
+  if (document.HasMember("dim_names")) {
+    for (auto& x : document["dim_names"].GetArray()) {
+      dim_names.emplace_back(x.GetString());
+    }
+    if (dim_names.size() != ndim) {
+      return Status::Invalid("Invalid dim_names");
+    }
+  }
+
+  std::vector<std::optional<int64_t>> uniform_shape;
+  if (document.HasMember("uniform_shape")) {
+    for (auto& x : document["uniform_shape"].GetArray()) {
+      if (x.IsNull()) {
+        uniform_shape.emplace_back(std::nullopt);
+      } else {
+        uniform_shape.emplace_back(x.GetInt64());
+      }
+    }
+    if (uniform_shape.size() != ndim) {
+      return Status::Invalid("uniform_shape size must match ndim. Expected: ", ndim,
+                             " Got: ", uniform_shape.size());
+    }
+  }
+
+  return variable_shape_tensor(value_type, static_cast<int32_t>(ndim), permutation,
+                               dim_names, uniform_shape);
+}
+
+std::shared_ptr<Array> VariableShapeTensorType::MakeArray(
+    std::shared_ptr<ArrayData> data) const {
+  DCHECK_EQ(data->type->id(), Type::EXTENSION);
+  DCHECK_EQ("arrow.variable_shape_tensor",
+            internal::checked_cast<const ExtensionType&>(*data->type).extension_name());
+  return std::make_shared<ExtensionArray>(data);
+}
+
+Result<std::shared_ptr<Tensor>> VariableShapeTensorType::MakeTensor(
+    const std::shared_ptr<ExtensionScalar>& scalar) {
+  const auto ext_scalar = internal::checked_pointer_cast<ExtensionScalar>(scalar);
+  const auto tensor_scalar = internal::checked_pointer_cast<StructScalar>(scalar->value);
+  const auto ext_type =
+      internal::checked_pointer_cast<VariableShapeTensorType>(scalar->type);
+
+  ARROW_ASSIGN_OR_RAISE(const auto shape_scalar, tensor_scalar->field(0));
+  ARROW_ASSIGN_OR_RAISE(const auto data_scalar, tensor_scalar->field(1));
+  const auto shape_array = internal::checked_pointer_cast<Int32Array>(
+      internal::checked_pointer_cast<FixedSizeListScalar>(shape_scalar)->value);
+  const auto data_array =
+      internal::checked_pointer_cast<BaseListScalar>(data_scalar)->value;
+
+  const auto value_type =
+      internal::checked_pointer_cast<FixedWidthType>(ext_type->value_type());
+
+  if (!is_fixed_width(*value_type)) {
+    return Status::Invalid("Cannot convert non-fixed-width values to Tensor.");
+  }
```

Comment:
We already know that the Scalar's type is VariableShapeTensorType, so we shouldn't need to do any more type checks here?
Please move extension type validation inside `VariableShapeTensorType::Make`.

Reply:
Indeed. Moved.